libcaf_net/caf/net/web_socket/framing.cpp

```diff
@@ -10,6 +10,49 @@
 
 namespace caf::net::web_socket {
 
+// -- static utility functions -------------------------------------------------
+
+expected<std::pair<uint16_t, const_byte_span>>
+framing::decode_closing_payload(const_byte_span payload) {
+  if (payload.size() == 0)
+    return std::make_pair(0, payload);
+  if (payload.size() < 2)
+    return make_error(caf::sec::protocol_error,
+                      "no status flag in closing payload");
+  auto status = (std::to_integer<uint_fast16_t>(payload[0]) << 8)
+                + std::to_integer<uint16_t>(payload[1]);
+  if (!detail::rfc3629::valid(payload.subspan(2)))
+    return make_error(sec::protocol_error,
+                      "malformed text message in closing payload");
+  if (status < 1000 || status >= 5000)
+    return make_error(sec::protocol_error,
+                      "invalid status code in closing payload");
+  // statuses between 3000 and 4999 are allowed and application specific
+  if (status >= 3000 && status < 5000)
+    return std::make_pair(status, payload.subspan(2));
+  // statuses between 1000 and 2999 are protocol defined
+  web_socket::status status_code;
+  auto success = from_integer(status, status_code);
+  if (!success)
+    return make_error(sec::protocol_error,
+                      "invalid status code in closing payload");
+  switch (status_code) {
+    case status::normal_close:
+    case status::going_away:
+    case status::protocol_error:
+    case status::invalid_data:
+    case status::inconsistent_data:
+    case status::policy_violation:
+    case status::message_too_big:
+    case status::missing_extensions:
+    case status::unexpected_condition:
+      return std::make_pair(status, payload.subspan(2));
+    default:
+      return make_error(sec::protocol_error,
+                        "invalid status code in closing payload");
+  }
```

Comment:
When looking at https://datatracker.ietf.org/doc/html/rfc6455#section-7.4.2, don't we actually have to accept everything in range 1000 to 4999? That would make the code just look like this:

```suggestion
  if (status < 1000 || status >= 5000)
    return make_error(sec::protocol_error,
                      "invalid status code in closing payload");
  return std::make_pair(status, payload.subspan(2));
```

Or am I missing something here?

Reply:
Status codes in the range 1000-2999 are reserved for definition by the WebSocket protocol. At [section 7.4.1](https://datatracker.ietf.org/doc/html/rfc6455#section-7.4.1) we see that only a limited set are defined at this moment and we should reject the other values. Even from the defined status code values, 1004 is undefined, and 1005 and 1006 must not be set as close control by definition, so we need to rule these out.